server/controller/plan.controller.js

```diff
@@ -0,0 +1,209 @@
+      return res.status(400).json({ error: SEMESTERS_FIELD_IS_REQUIRED });
+    } else if (req.body.semesters.constructor !== Object) {
+      return res.status(400).json({ error: SEMESTERS_IS_INVALID });
+    }
+
+    try {
+      const { user_id, semesters } = req.body;
+
+      // insertMany() documentation: https://docs.mongodb.com/manual/reference/method/db.collection.insertMany/
+      /*
+        const insertManyResult = await insertMany(semesters);
+        const insertOneResult = await insertOne({ user_id, insertManyResult.ids });
+      */
+
+      return res.status(200).json();
+    } catch (databaseError) {
+      return res.status(500).json({ error: databaseError });
+    }
+  }
+);
+
+/**
+ * @route   PUT /plan/update
+ * @desc    Updates degree plans
+ * @access  Private
+ */
+planController.put(
+  '/update',
+  passport.authenticate('jwt', { session: false }),
+  validateUpdatePlan,
+  async (req, res) => {
+    // Can't test this because there is no semester schema yet
+    // But I am assuming that the client will send the user's id as well as an array of objects each containing all fields of a semester document
+    // something like this:
+    /*
+      {
+        "user_id": "a966a6a66ba6b8a698b6",
+        "semesters": [
+          "courses": [Courses.ID],
+          ...,
+          ...
+        ]
+      }
+    */
+
+    const errors = validationResult(req);
+    if (errors.isEmpty() === false) {
+      return res.status(400).json(errors);
+    }
+
+    const { user_id, plan_id, semesters } = req.body;
+
+    if (!user_id && !plan_id) {
+      return res.status(400).json({ error: USER_OR_PLAN_ID_REQUIRED });
+    } else if (!semesters) {
+      // Temporary validation for semesters. Requires improvement upon addition of Semester schema
+      return res.status(400).json({ error: SEMESTERS_FIELD_IS_REQUIRED });
+    } else if (semesters.constructor !== Object) {
+      return res.status(400).json({ error: SEMESTERS_IS_INVALID });
+    }
+
+    try {
+      /*
+        Either update via user_id or plan_id
+        Determine which semesters in the array require updating
+        update the needed semesters
+      */
+      return res.status(200).json();
+    } catch (databaseErrors) {
+      return res.status(500).json({ error: databaseErrors });
+    }
+  }
+);
+
+/**
+ * @route   GET /plan/
+ * @desc    Fetches degree plans
+ * @access  Private
+ */
+planController.get(
+  '/',
+  passport.authenticate('jwt', { session: false }),
+  validateFetchPlan,
+  async (req, res) => {
+    const errors = validationResult(req);
+    if (errors.isEmpty() === false) {
+      return res.status(400).json(errors);
+    } else if (!req.query.plan_id && !req.query.user_id) {
+      return res.status(400).json({ error: USER_OR_PLAN_ID_REQUIRED });
+    }
+
+    try {
+      const { plan_id, user_id, fetch_data } = req.query;
+
+      const fetchedPlan = await Plan.findOne({
+        _id: user_id ? user_id : plan_id,
```

Comment:
I'm removing the user_id field in the schema for Plan. The code for all requests are going to be changed to remove the user_id field in the schema. However, for this request, I'm thinking it would make the Plan API more flexible by allowing the user to request a plan by the user's _id or by the plan's _id. If by the user's _id then we'd have to make 2 queries, if by the plan_id then only one.